### PR TITLE
Improvements to the LST workflows and small general LST fixes

### PR DESCRIPTION
**Pull request by VourMa**
This PR follows up on some items from #46746. More specifically, I have:
- Moved the LST workflows to use the latest default geometry (D110)
- Made the `0.704` run on a GPU only if there is one, otherwise fall back to CPU (hardware-agnostic workflow, in accordance with the rest of heterogeneous workflows in CMSSW).
- Replaced the custom binary search function with `std::lower_bound`.
- Got rid of the warning about the `SeedStopInfo` size mismatch between the `TrackCandidate` and the `TrajectorySeed` collections.
- Made a couple more minor fixes following up on PR integration comments (e.g. using `MessageLogger` instead of `printf`, renamings, etc.)

I have tested the changes privately and the performance is exactly the same, so this is a technical PR.

### CODE REVIEW

**Comment by slava77** [comment]
test parameters:
- enable_tests = gpu
- workflows = 29634.703,29834.703
- workflows_gpu = 29634.704,29834.704
- relvals_opt = -w upgrade
- relvals_opt_gpu = -w upgrade

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@smuzaffar 
are the CPU comparisons missing due to failing DAS queries? Perhaps these can still be allowed to run for the passing tests.
Please check.
Thank you.

**Comment by smuzaffar** [comment]
> are the CPU comparisons missing due to failing DAS queries? 

yes @slava77 , comparison job is not run if there are relval failures. 

> Perhaps these can still be allowed to run for the passing tests.

sure, sounds reasonable. We will update bot to run comparison job for passed relvals

**Comment by slava77** [comment]
> are the CPU comparisons missing due to failing DAS queries?

the question is somewhat academic for this PR though: the workflow numbers in this PR are not present in the baseline and the affected code does not run in other requested workflows. 
So, changes in the code can not be tested in the bot comparisons (only that the code runs).

**Comment by smuzaffar** [comment]
For valid additioanl workflows, bot does try to generate baseline. For this PR it did try to generate baseline for `29634.703,29834.703`  but failed as these workflows are not available at least `runTheMatrix.py -n -w upgrade -l 29634.703,29834.703` failed https://cmssdt.cern.ch/jenkins/job/ib-run-baseline/13661/console

**Comment by slava77** [comment]
> Note, that bot can only generate baselines for workflows if they are available in IB. Any new workflow added by PR will not have any baseline

Yes, that's what I implied in my ["academic" comment](https://github.com/cms-sw/cmssw/pull/46967#issuecomment-2548666719)

Actually, only one workflow is not defined in the baseline (`29834`), let me update the test parameters to just 634, also hoping the DAS errors go away

**Comment by slava77** [comment]
test parameters:
- enable_tests = gpu
- workflows = 29634.703
- workflows_gpu = 29634.704
- relvals_opt = -w upgrade
- relvals_opt_gpu = -w upgrade

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
FractionOfCandOverSeeds_highPtTripletStepTrackCandidates_highPtTripletStep changed in 29634.703
image

I guess it's from https://github.com/cms-sw/cmssw/pull/46967/files#diff-a67b503ef4abafb8f3149bdb2412b9e125b2b0de50f969d52c5400e8267823aaR398-R402
`TrackSeedMonhighPtTripletStep.SeedProducer = "lstPixelSeedInputProducer"`
@VourMa do you agree?

**Comment by VourMa** [comment]
> FractionOfCandOverSeeds_highPtTripletStepTrackCandidates_highPtTripletStep changed in 29634.703 image
> 
> I guess it's from https://github.com/cms-sw/cmssw/pull/46967/files#diff-a67b503ef4abafb8f3149bdb2412b9e125b2b0de50f969d52c5400e8267823aaR398-R402 `TrackSeedMonhighPtTripletStep.SeedProducer = "lstPixelSeedInputProducer"` @VourMa do you agree?

Yes

**Comment by slava77** [comment]
(after 2 days of no indication of review)
@cms-sw/dqm-l2 @cms-sw/reconstruction-l2 @cms-sw/upgrade-l2 @cms-sw/pdmv-l2 
please note that tests in this PR have completed and show expected performance.
Please take a look and comment or perhaps sign.
Thank you.

**Comment by jfernan2** [comment]
+1

**Comment by Moanwar** [comment]
+Upgrade

**Comment by AdrianoDee** [comment]
+pdmv

File: Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py [hunk 1]
```diff
@@ -498,8 +498,9 @@ def setup__(self, step, stepName, stepDict, k, properties):
         if 'Reco' in step: stepDict[stepName][k] = merge([self.step3, stepDict[step][k]])
         elif 'HARVEST' in step: stepDict[stepName][k] = merge([{'-s': 'HARVESTING:@trackingOnlyValidation+@trackingOnlyDQM'}, stepDict[step][k]])
         elif 'ALCA' in step: stepDict[stepName][k] = None
-    def condition_(self, fragment, stepList, key, hasHarvest):
-        return ('Run4' in key)
+    def condition(self, fragment, stepList, key, hasHarvest):
+        result = (fragment=="TTbar_14TeV") and hasHarvest and ('Run4' in key)
```
Comment (AdrianoDee) [inline Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py:502]: Just for my curiosity (already signed): why limiting this to `TTbar`? Could be useful to have it defined for some displaced topologies?

**Comment by antoniovagnerini** [comment]
+dqm
- spurious differences from HGCal DQM test (open issue  #46416)

File: Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py (hunk 1 above)
Comment (VourMa) [reply Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py:502]: The workflow was built to inherit from `UpgradeWorkflowTracking`, so what I did here was to override this line:
https://github.com/cms-sw/cmssw/blob/6c309578b4970a207343f7ae56a6decd242bdb8e/Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py#L298
to include the one sample of Phase 2 and to remove `not PU` restriction.
I agree that more workflows would be useful but there will be more work on our workflows in the future, which will address this point as well.

**Comment by mandrenguyen** [comment]
+1

File: Configuration/PyReleaseValidation/python/relval_gpu.py [hunk 2]
```diff
@@ -70,13 +70,11 @@
            13061.402,
 
            # Run4, Alpaka-based noPU
-           29634.402, 29634.403, 29634.404, 29634.406,
+           29634.402, 29634.403, 29634.404, 29634.406, 29634.704,
```
[the remark this answers is not in the document]
Comment (VourMa) [reply Configuration/PyReleaseValidation/python/relval_gpu.py:73]: Sure, sorry for missing it. I will prepare a PR to fix this.

**Pull request merged**